apollo-integration/src/test/java/com/apollographql/apollo/ApolloWatcherTest.java

```diff
@@ -104,6 +111,50 @@ public void testQueryWatcherUpdated_SameQuery_DifferentResults() throws IOExcept
   }
 
   @Test
+  public void testQueryWatcherUpdated_Store_write() throws IOException, InterruptedException,
+      TimeoutException, ApolloException {
+    final NamedCountDownLatch firstResponseLatch = new NamedCountDownLatch("firstResponseLatch", 1);
+    final NamedCountDownLatch secondResponseLatch = new NamedCountDownLatch("secondResponseLatch", 2);
+
+    EpisodeHeroNameQuery query = EpisodeHeroNameQuery.builder().episode(Episode.EMPIRE).build();
+    server.enqueue(mockResponse("EpisodeHeroNameResponseWithId.json"));
+
+    ApolloQueryWatcher<EpisodeHeroNameQuery.Data> watcher = apolloClient.query(query).watcher();
+    watcher.enqueueAndWatch(
+        new ApolloCall.Callback<EpisodeHeroNameQuery.Data>() {
+          @Override public void onResponse(@Nonnull Response<EpisodeHeroNameQuery.Data> response) {
+            if (secondResponseLatch.getCount() == 2) {
+              assertThat(response.data().hero().name()).isEqualTo("R2-D2");
```

Comment:
> We still need to check that so my favorite solution would be to have the callback run in the main thread

Yes we have to fix other places as well, and we have a ticket for that

> Is it ok to use AndroidSchedulers.mainThread() in there or are we pure java ?

It's a pure java test. What you can do is use `AtomicReference`  to set all what you need inside callback, and assert then outside callback. (Ugly I know but not sure what else we can do)